src/gui_common/AddWindowReorderingSupportToSiblings.cs

```diff
@@ -0,0 +1,328 @@
+                    window, windowReorderingAncestors[i].Sibling, recursive);
+            }
+        }
+
+        window.Connect(nameof(Dragged), this, nameof(OnWindowReorder));
+
+        connectedWindows.Add(window, topNode);
+        connectedSiblings.Add(topNode);
+
+        // Update top sibling
+        if (topSibling is null || topSibling.GetIndex() < topNode.GetIndex())
+            topSibling = topNode;
+    }
+
+    public void DisconnectWindow(CustomWindow window, bool recursive)
+    {
+        if (recursive)
+        {
+            // Connection was recursive, disconnect from other window reordering nodes
+            for (int i = 0; i < windowReorderingAncestors.Count; i++)
+            {
+                windowReorderingAncestors[i].Node.DisconnectWindow(window, recursive);
+            }
+        }
+
+        window.Disconnect(nameof(Dragged), this, nameof(OnWindowReorder));
+
+        var windowSibling = connectedWindows[window];
+        connectedWindows.Remove(window);
+
+        // Find if another connected window has the same sibling
+        bool foundSibling = false;
+        foreach (var sibling in connectedSiblings)
+        {
+            if (sibling == windowSibling)
+            {
+                foundSibling = true;
+                break;
+            }
+        }
+
+        if (!foundSibling)
+        {
+            // No other window has the same sibling so it can be removed
+            connectedSiblings.Remove(windowSibling);
+        }
+    }
+
+    /// <summary>
+    ///   Finds whose windows ancestor is currently on the top
+    /// </summary>
+    public void UpdateTopSibling()
+    {
+        if (topSibling is not null)
+        {
+            // Top sibling is already current
+            return;
+        }
+
+        var siblings = parent.GetChildren();
+
+        // Search for the first sibling that is part of the reordering system
+        for (int i = siblings.Count - 1; i >= 0; i--)
+        {
+            var sibling = (Node)siblings[i];
+            if (!connectedSiblings.Contains(sibling))
+                continue;
+
+            topSibling = sibling;
+            break;
+        }
+    }
+
+    /// <summary>
+    ///   Reoders a window by setting its ancestor to the position of current top window, making it appear on top of
+    ///   others.
+    /// </summary>
+    public void OnWindowReorder(CustomDialog window)
+    {
+        // Get a sibling that is an ancestor of this window
+        Node targetSibling = connectedWindows[window];
+
+        UpdateTopSibling();
+
+        int topSiblingIndex = topSibling!.GetIndex();
+
+        if (topSiblingIndex == targetSibling.GetIndex())
+        {
+            // This window is already on the top
+            return;
+        }
+
+        // Put the sibling on the top
+        parent.MoveChild(targetSibling, topSiblingIndex);
+
+        topSibling = targetSibling;
+
+        // For unexplained reasons this has to be here to update the order visually
```

Comment:
Still sounds a bit like a hack. One potential bug I realized is that this doesn't preserve the on top state. Instead as this hack is required for now the current top level should be stored and restored like:
```c#
bool isTopLevel = window.IsTopLevel(); // I hope there's a method or property like this
window.SetTopLevel(!isTopLevel);
window.SetTopLevel(isTopLevel);
```

At least that code will leave the window in the same state as it previously was. Because otherwise I don't want to have to debug the new popup modal system that has already had an absolute ton of bugs related to it.

Reply:
For some reason I assumed that windows are always on the top. I wonder if this hack is necessary for windows that are not on the top too. But it will probably be better to use it for them too just in case... or to do more testing to know if it's really necessary.

Reply:
Only modal popups (that disallow you from interacting with the rest of the GUI) are set as top level. The relevant code is: 

https://github.com/Revolutionary-Games/Thrive/blob/f0c0322add610681556122cae3687519fefd37b7/src/gui_common/ModalManager.cs#L10

Reply:
I don't quite fully get why this is necessary. As I understand it, the visual order of the windows should already be updated when their position is changed in the tree, at least that's what I observed in `ModalManager`.

CustomWindows and its descendants should always be top level, I don't see why they would otherwise?

Reply:
Oh sorry, I mixed up being top level and being modal. Like usual, that PR where this was an important distinction still haunts me it seems.

But yeah, I also found this very strange which is why I said that if this code is kept we need to open an issue about trying to remove the hack at some point if we can't come up with anything for this PR.

Reply:
I opened the issue about this hack: https://github.com/Revolutionary-Games/Thrive/issues/4349